rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,512 @@
+---
+authors: Trent Clarke (trent@goteleport.com)
+state: draft
+---
+
+# RFD 2 - Custom Terraform Registry
+
+## What
+
+A Terraform _registry_ is an index of available terraform _providers_. The
+Terraform tool uses registries to locate, validate and install Terraform
+providers for the various services required by terraform scripts. Hashicorp
+(the Terraform vendor) publishes the protocol that Terraform uses to interact
+with registries and so third parties (like us) may provide _custom_ Terraform
+registries outside the main, Hashicorp-provided public registry.
+
+This RFD investigates the use of a custom Terraform registry to distribute the
+Teleport provider for Terraform.
+
+## Why
+
+We have two main goals that effect how we offer our Terraform provider to
+customers
+
+ 1. We want to meet out customers where they are, e.g. if they're using
+    Terraform, we want to supply them with a Terraform provider, and a method
+    for installing it that is natural to the Terraform ecosystem. This is a
+    good practice in general and, in the specific, is part of our Q1 focus on
+    improving our Time-to_first-Value.
+
+ 2. We want to maintain close control of the distribution of our software,
+    including supply chain management and monitoring of publishing, changes,
+    downloads and host of other activities.
+
+With these two goals in mind, the most practical solution for distributing
+the Teleport Terraform provider appears to be hosting a custom Terraform
+registry.
+
+Note that the concerns in point 2 are being addressed in a more general scope
+by [Cloud RFD 0004](https://github.com/gravitational/cloud/blob/master/rfd/0004-Release-Asset-Management.md),
+and the solution proposed by this document will endeavour to be compatible
+Cloud RFD 004 as far as possible.
+
+### Why not use the public Hashicorp registry?
+
+This conflicts with goal #2 (i.e. maintaining close control of the
+distribution of our software).
+
+### Will using a custom registry be onerous on our users?
+
+No. Certainly less onerous than the current solution (involving several
+manual download and installation steps).
+
+The user would reference the custom registry directly in their terraform
+scripts, and the Terraform tool would take care of the rest.
+
+```terraform
+terraform {
+    required_providers {
+        teleport = {
+            source = "terraform.releases.teleport.dev/gravitational/teleport"
+            version = "9.1.0"
+        }
+    }
+}
+```
+
+For more info, see [the Terraform docs](https://www.terraform.io/cloud-docs/registry/using)
+on using custom registries.
+
+### What about name squatting on the Hashicorp registry?
+
+What to do about people publishing homemade Teleport providers to the
+Hashicorp registry is beyond the scope of this RFD.
+
+It's worth noting, though, that _not_ being on the main Hashicorp registry _does_
+make our official solution harder to find.
```

Comment:
This is one of my biggest issues with us looking to use a custom registry instead of the one that basically everyone else does. People are going to go there, search for Teleport and not find us. If they Google, they probably will find us, but some will definitely think "why on earth aren't they using the main registry?"

Reply:
👍 -- I'm not seeing any significant value yet where having a private registry would be better, especially for the overhead involved.

Reply:
@webvictim @reedloden @tcsc Trent wrote this RFD following my ask to have all assets that we distribute to be under our control, including Terraform registry. That's primary reason for us hosting private registry, similarly to our parallel efforts of hosting our all assets. The primary reason for that is to protect us from supply chain attacks. Having all assets distributed by teleport CDN is a costly initial investment, but a year ago we made this trade off to make sure we will not be vulnerable to these class of attacks.

I'm open to revisiting this discussion and providing other type of guidance to developers, but let's do this a bit later and outside of this PR, as this is a larger discussion.

Reply:
Good to know.

Side note, are we tracking https://github.com/gravitational/teleport/issues/4870 in that same vein? We're already offering Homebrew as an option on our [install guide](https://goteleport.com/docs/getting-started/linux-server/#install-a-teleport-client-locally), and [analytics](https://formulae.brew.sh/formula/teleport#default) show it is in wide use.